cf-oscore/src/test/java/org/eclipse/californium/oscore/ContextRederivationTest.java

```diff
@@ -0,0 +1,127 @@
+	private final static String uriLocal = "coap://localhost";
+	private final static String hello1 = "/hello/1";
+	private final static AlgorithmID alg = AlgorithmID.AES_CCM_16_64_128;
+	private final static AlgorithmID kdf = AlgorithmID.HKDF_HMAC_SHA_256;
+
+	// test vector OSCORE draft Appendix C.1.1
+	private final static byte[] master_secret = { 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B,
+			0x0C, 0x0D, 0x0E, 0x0F, 0x10 };
+	private final static byte[] master_salt = { (byte) 0x9e, (byte) 0x7c, (byte) 0xa9, (byte) 0x22, (byte) 0x23,
+			(byte) 0x78, (byte) 0x63, (byte) 0x40 };
+	private final static byte[] sid = new byte[0];
+	private final static byte[] rid = new byte[] { 0x01 };
+	
+	/**
+	 * Test context re-derivation followed by a normal message exchange.
+	 * 
+	 * @throws OSException
+	 * @throws ConnectorException
+	 * @throws IOException
+	 */
+	@Test
+	public void rederivationTest() throws OSException, ConnectorException, IOException {
+		OSCoreCtx ctx = new OSCoreCtx(master_secret, true, alg, sid, rid, kdf, 32, master_salt, null);
+		db.addContext(uriLocal, ctx);
+		OSCoreCoapStackFactory.useAsDefault();
+
+		rederive(uriLocal);
+		
+		CoapClient c = new CoapClient(uriLocal + hello1);
+		Request r = new Request(Code.GET);
+		r.getOptions().setOscore(new byte[0]);
+		System.out.println((Utils.prettyPrint(r)));
+		
+		CoapResponse resp = c.advanced(r);
+		System.out.println((Utils.prettyPrint(resp)));
+		
+		assertEquals(resp.getCode(), ResponseCode.CONTENT);
+		assertEquals(resp.getResponseText(), SERVER_RESPONSE);
+	}
+
+	/**
+	 * Perform re-derivation of contexts as detailed in Appendix B.2.
+	 * Essentially it uses a message exchange together with the Context ID
+	 * field in the OSCORE option to securely generate a new shared context.
+	 * 
+	 * @throws IOException 
+	 * @throws ConnectorException 
+	 * @throws OSException 
+	 */
+	public static void rederive(String uriLocal) throws ConnectorException, IOException, OSException {
+		//Generate a random 8 byte Context ID
+		SecureRandom random = new SecureRandom();
+		byte[] newContextId = new byte[8];
+		random.nextBytes(newContextId);
+		
+		//Create new context with the generated Context ID
+		OSCoreCtx ctx = new OSCoreCtx(master_secret, true, alg, sid, rid, kdf, 32, master_salt, newContextId);
+		ctx.setIncludeContextId(true);
+		db.addContext(uriLocal, ctx);
+		OSCoreCoapStackFactory.useAsDefault();
+		
+		//Now send request using the new context
+		String resource = "/rederive"; //Dummy resource to access for context re-derivation
+		String URI = uriLocal + resource;
+		System.out.println(URI);
+		OSCoreCoapStackFactory.useAsDefault();
+		
+		CoapClient c = new CoapClient(URI);
+		Request r = new Request(Code.GET);
+		r.getOptions().setOscore(new byte[0]);
+		System.out.println(Utils.prettyPrint(r));
+		
+		CoapResponse resp = null;
+		resp = c.advanced(r);
+		System.out.println(Utils.prettyPrint(resp));
```

Comment:
If I run the unit test, it seems to hang here and then report a NullPointerException

Reply:
Yes, currently the HelloWorldServer must be started manually before this JUnit test works to run independently. I will fix this.

Reply:
OK. If it's intended to be merged before that is fixed, the test must be marked with @Ignore, otherwise all ci builds will fail.

Reply:
Yes, I have now done this in commit cdecae7c41deeb2a8f26ee011fe6747a97f4acf0

Then I can fix it in the next PR to finalize support for context re-derivation.